src/midi/midi_fluidsynth.cpp

```diff
@@ -682,33 +680,36 @@ void MidiHandlerFluidsynth::RenderAudioFramesToFifo(const uint16_t num_audio_fra
 void MidiHandlerFluidsynth::ProcessWorkFromFifo()
 {
 	const auto work = work_fifo.Dequeue();
+	if (!work) {
+		return;
+	}
 
 	/* // Comment-in to log inter-cycle rendering
-	if ( work.num_pending_audio_frames > 0) {
+	if ( work->num_pending_audio_frames > 0) {
 	        LOG_MSG("FSYNTH: %2u audio frames prior to %s message, followed
 	by "
 	                "%2lu more messages. Have %4lu audio frames queued",
-	                work.num_pending_audio_frames,
-	                work.message_type == MessageType::Channel ? "channel" :
+	                work->num_pending_audio_frames,
+	                work->message_type == MessageType::Channel ? "channel" :
 	"sysex", work_fifo.Size(), audio_frame_fifo.Size());
 	}*/
 
-	if (work.num_pending_audio_frames > 0) {
-		RenderAudioFramesToFifo(work.num_pending_audio_frames);
+	if (work->num_pending_audio_frames > 0) {
+		RenderAudioFramesToFifo(work->num_pending_audio_frames);
 	}
 
-	if (work.message_type == MessageType::Channel) {
-		ApplyChannelMessage(work.message);
+	if (work->message_type == MessageType::Channel) {
+		ApplyChannelMessage(work->message);
 	} else {
-		assert(work.message_type == MessageType::SysEx);
-		ApplySysexMessage(work.message);
+		assert(work->message_type == MessageType::SysEx);
+		ApplySysexMessage(work->message);
 	}
 }
 
 // Keep the fifo populated with freshly rendered buffers
 void MidiHandlerFluidsynth::Render()
 {
-	while (keep_rendering.load()) {
+	while (!work_fifo.StoppedEnqueuing()) {
```

Comment:
Yeah, so (to me) `while (work_fifo.IsRunning())` would be perhaps a bit more natural than the "double negation".

Reply:
definitely agree; will do.

Reply:
The queue represents a form of defered work.
 - Stopping it immediate prevents new work from entering.
 - But a choice needs (or can) be made about the current work in the queue.
 - Do you ignore that work or complete it?

In the case of audio samples - we don't care and just throw the queue away.. the audio stream just comes to a close sooner.

When it comes to important stuff in the queue, it makes sense to finish that up.

So that's why we've got two indicators.

Thinking we can flip them to the positive.
 - IsEnqueuing() -> when false, queuing is futile and discarded.
 - IsDequeuing() -> when false, the queue has run dry /and/ there won't be more.
 -

Reply:
A general option might be more readable- IsRunning(), could be IsEnqueuing() || IsDequeuing() under the hood.

Reply:
> * IsEnqueuing() -> when false, queuing is futile and discarded.
> 
> * IsDequeuing() -> when false, the queue has run dry /and/ there won't be more.

Yeah I like that. Better than inferring it from IsRunning() and IsEmpty(), on a second thought.